src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpRetry.cs

```diff
@@ -0,0 +1,248 @@
+        return Convert.ToInt32(nextMilliseconds);
+    }
+
+    private static TimeSpan? TryGetGrpcRetryDelay(StatusCode statusCode, Metadata trailers)
+    {
+        Debug.Assert(trailers != null, "trailers was null");
+
+        if (statusCode != StatusCode.ResourceExhausted && statusCode != StatusCode.Unavailable)
+        {
+            return null;
+        }
+
+        var statusDetails = trailers.Get(GrpcStatusDetailsHeader);
+        if (statusDetails != null && statusDetails.IsBinary)
+        {
+            var status = Status.Parser.ParseFrom(statusDetails.ValueBytes);
+            foreach (var item in status.Details)
+            {
+                var success = item.TryUnpack<RetryInfo>(out var retryInfo);
+                if (success)
+                {
+                    return retryInfo.RetryDelay.ToTimeSpan();
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private static TimeSpan? TryGetHttpRetryDelay(HttpStatusCode statusCode, HttpResponseHeaders headers)
+    {
+        Debug.Assert(headers != null, "headers was null");
+
+#if NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
+        return statusCode == HttpStatusCode.TooManyRequests || statusCode == HttpStatusCode.ServiceUnavailable
+#else
+        return statusCode == (HttpStatusCode)429 || statusCode == HttpStatusCode.ServiceUnavailable
+#endif
+            ? headers.RetryAfter?.Delta
+            : null;
+    }
+
+    private static bool IsGrpcStatusCodeRetryable(StatusCode statusCode, bool hasRetryDelay)
+    {
+        switch (statusCode)
+        {
+            case StatusCode.Cancelled:
+            case StatusCode.DeadlineExceeded:
+            case StatusCode.Aborted:
+            case StatusCode.OutOfRange:
+            case StatusCode.Unavailable:
+            case StatusCode.DataLoss:
+                return true;
+            case StatusCode.ResourceExhausted:
+                return hasRetryDelay;
+            default:
+                return false;
+        }
+    }
+
+    private static bool IsHttpStatusCodeRetryable(HttpStatusCode statusCode, bool hasRetryDelay)
+    {
+        switch (statusCode)
+        {
+#if NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
+            case HttpStatusCode.TooManyRequests:
+#else
+            case (HttpStatusCode)429:
+#endif
+            case HttpStatusCode.BadGateway:
+            case HttpStatusCode.ServiceUnavailable:
+            case HttpStatusCode.GatewayTimeout:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    private static int GetRandomNumber(int min, int max)
+    {
+#if NET6_0_OR_GREATER
+        return Random.Shared.Next(min, max);
+#else
+        // TODO: Implement this better to minimize lock contention.
+        // Consider pulling in Random.Shared implementation.
+        lock (Random)
+        {
+            return Random.Next(min, max);
```

Comment:
Do we want to use `max + 1`? The second argument for `Random.Next` is the exclusive upper bound of the random number returned.

Reply:
🤷 Not sure it matters a whole lot. I'm happy to go either way. Just for some additional context, the behavior of the retry logic in this PR is very much inspired by grpc-dotnet's implementation of the gRPC retry specification including the calculation for random jitter.

Using `Random.Next(min, max)` is equivalent to grpc-dotnet's implementation

https://github.com/grpc/grpc-dotnet/blob/01049834beb8342d1cf93bc5902a7fb2d3a310e1/src/Grpc.Net.Client/Internal/Retry/RetryCall.cs#L216

Reply:
> Not sure it matters a whole lot.

Probably not. It's more of a code readability thing. It might seem a bit odd as the variable names suggest that the max value could indeed be a possible `delayDuration` value.

I'm okay with the code as-is.